torch/optim/lr_scheduler.py

```diff
@@ -21,10 +22,11 @@
 
 SAVE_STATE_WARNING = "Please also save or load the state of the optimizer when saving or loading the scheduler."
 
-class _LRScheduler(object):
 
-    def __init__(self, optimizer, last_epoch=-1, verbose=False):
+class _LRScheduler(object):
 
+    def __init__(self, optimizer: Optimizer, last_epoch: int = -1,
```

Comment:
there are many internal pyre errors `Incompatible parameter type`, e.g.  `Expected torch.optim.optimizer.Optimizer` for 1st positional only parameter to call `torch.optim.lr_scheduler.ReduceLROnPlateau.__init__` but got `torch.optim.Optimizer`.`

Reply:
That's the same class and the errors seem to be a clear Pyre bug/limitation. I cannot find an open Pyre issue for it though.

Changing the imports to `from . import Optimizer` seems wrong, that can lead to import cycles. I'd be inclined to add `# pyre-ignore` to all the `optimizer: Optimizer` annotations and report this as a Pyre bug. Would that be okay?

Reply:
@ezyang -- have you encountered this?

Reply:
I spent some time on trying to reproduce this, but was unable to do so with the latest `pyre` release (`0.0.56`). Perhaps it's the combination of the `del optimizer` in `optim/__init__.py` with a complicated code base that makes `pyre` unable to detect that `torch.optim.Optimizer` and `torch.optim.optimizer.Optimizer` are the same thing.

I believe `torch.optim.optimizer` and other filenames that are `del`'ed in `__init__.py` are not public API, so my next step if I'd have to try to fix this without knowing what is wrong is to remove `optimizer.py` to `_optimizer.py` and get rid of all the `del` statements. 

@vincentqb any chance you can give me a reproducer?

Reply:
@vincentqb I've only had some situations where I broke Pyre for downstream projects; check out https://www.internalfb.com/intern/tasks/?t=65894941

If it's internal to our project we should just internally figure out a solution one way or another. If you don't have time to figure it out we should find someone else internally to handle it.

Reply:
> @vincentqb I've only had some situations where I broke Pyre for downstream projects; check out https://www.internalfb.com/intern/tasks/?t=65894941
> 
> If it's internal to our project we should just internally figure out a solution one way or another. If you don't have time to figure it out we should find someone else internally to handle it.

It's on a downstream project, see [phabricator](https://www.internalfb.com/intern/diff/D23165271/). Who do you suggest pinging for this? That would help, thanks!

Reply:
Vincent, there is some basic due diligence you should do first. Because fbcode is a monorepo, it is pretty easy to try, e.g., Ralf's suggestion at https://github.com/pytorch/pytorch/pull/43089#discussion_r499064708 and then rerun the Pyre job as per the instructions in Sandcastle (`pyre -l oculus/eye`)

It would be appropriate to ping the downstream team if you believe that there is a bug in their annotation. I looked at their code and it looks correct to me:

```
  @timer
  def create_scheduler(
      logger,
      optimizer: torch.optim.Optimizer,
      start_epoch: int,
      config: cfg.SchedulerConfig,
      # pyre-fixme[11]: Annotation `lr_scheduler` is not defined as a type.
  ) -> torch.optim.lr_scheduler:
      if config.schedulerType == "StepLR":
          lr_config = config.step_lr_config
          return torch.optim.lr_scheduler.StepLR(
              optimizer,
              step_size=lr_config.stepSize,
              gamma=lr_config.gamma,
              last_epoch=start_epoch - 1,
          )
```

so I don't think it's appropriate to reach out to them. You could probably bash this particular error by changing the type of optimizer but you'd probably have a similar problem crop up at use sites of `create_scheduler`. You could also bash out the problem by adding a `pyre-ignore` here, but that also feels wrong to me as there seem to be a lot of places where you would have to add pyre-ignore in this case.

Alternately, it's appropriate to reach out to Pyre team if you think there is a Pyre bug. Pyre has a workplace group at https://fb.workplace.com/groups/pyreqa It will be easier for them to help you if you have a minimal repro, as suggested by Ralf above. 

My preferred resolution is we change PyTorch itself in such a way that we no longer trigger this Pyre bug.

Reply:
@vincentqb please let me know if I can do anything to move this forward, or if it's stuck and needs a different plan. You kindly volunteered to land this (thank you), and probably didn't expect to have to deal with this kind of type checker issue.

Reply:
Thanks for the reminder, I'm looking into it.

Reply:
Rebased and applied change suggested [here](https://github.com/pytorch/pytorch/pull/43089#discussion_r499064708). The experiment in #46488 looks good so far, but let's keep an eye on it.

Reply:
The same issue comes up in #46488. I've opened an issue on [pyreqa](https://fb.workplace.com/groups/pyreqa/permalink/3481269998629474/).

Reply:
According to the post, and experiments on [the copy](https://github.com/pytorch/pytorch/pull/46488), there's an internal type files that need to be removed also (see [internal post](https://fb.workplace.com/groups/pyreqa/permalink/3481269998629474/) for more details). However, doing so opens quite a lot of type errors in the [internal diff](https://www.internalfb.com/intern/diff/D24370552/). They either were hidden by the internal file, or the fix is not the right one.

@izdeby -- since you've commandeered the diff, are you looking into the type errors in the [internal diff](https://www.internalfb.com/intern/diff/D24370552/)?